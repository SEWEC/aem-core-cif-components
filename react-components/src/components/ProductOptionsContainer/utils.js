/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

export function createBundleState(bundleItems) {
    return bundleItems.map(item => {
        return {
            option_id: item.option_id,
            required: item.required,
            title: item.title,
            type: item.type,
            quantity: ['checkbox', 'multi'].includes(item.type) ? 1 : item.options.find(o => o.is_default).quantity,
            options: item.options
                .slice() // return a shallow copy of the array. the original is frozen and cannot be sorted as Array.prototype.sort() sorts the elements of the array in place
                .sort((a, b) => a.position - b.position)
                .map(option => {
                    const { id, can_change_quantity, quantity, label, price } = option;

                    return {
                        id,
                        can_change_quantity,
                        quantity,
                        label,
                        price
                    };
                }),
            customization: item.options
                .filter(o => o.is_default)
                .map(o => {
                    const { id, quantity, price } = o;

                    return {
                        id,
                        quantity,
                        price
                    };
                })
        };
    });
}

export function getBundlePrice(bundleState) {
    return bundleState
        ? bundleState.reduce((acc, selection) => {
              return (
                  acc +
                  selection.quantity *
                      selection.customization.reduce((a, c) => {
                          return a + c.price * (['checkbox', 'multi'].includes(selection.type) ? c.quantity : 1);
                      }, 0)
              );
          }, 0)
        : 0;
}

export function getBundleCartOptions(bundleState) {
    if (bundleState) {
        return {
            bundle: true,
            options: bundleState.map(s => {
                return {
                    id: s.option_id,
                    quantity: s.quantity,
                    value: s.customization.map(c => c.id.toString())
                };
            })
        };
    }

    return {};
}

export function isOptionsStateValid(state) {
    return state
        ? state.reduce((acc, option) => {
              return acc && (option.required ? option.customization.length > 0 : true);
          }, true)
        : true;
}

export function createCustomizableOptionsState(options) {
    return options.map(item => {
        let result = {
            option_id: item.option_id,
            required: item.required,
            title: item.title,
            type: item.__typename
        };

        if (
            [
                'CustomizableDropDownOption',
                'CustomizableMultipleOption',
                'CustomizableRadioOption',
                'CustomizableCheckboxOption'
            ].includes(item.__typename)
        ) {
            result.options = item.value__multi
                .slice() // return a shallow copy of the array. the original is frozen and cannot be sorted as Array.prototype.sort() sorts the elements of the array in place
                .sort((a, b) => a.position - b.position)
                .map(o => {
                    return {
                        id: o.option_type_id,
                        label: o.title,
                        price: o.price
                    };
                });

            result.customization = item.required ? [result.options[0]] : [];
        } else {
            result.value = item.value;
            result.customization = '';
        }

        return result;
    });
}

export function getCustomizableOptionsPrice(optionsState) {
    return optionsState
        ? optionsState.reduce((acc, selection) => {
              const { value, customization } = selection;
              return (
                  acc +
                  (value
                      ? customization.length > 0
                          ? value.price
                          : 0
                      : customization.reduce((a, c) => a + c.price, 0))
              );
          }, 0)
        : 0;
}

export function getCustomizableCartOptions(optionsState) {
    if (optionsState) {
        const customizable_options = [];

        optionsState.forEach(selection => {
            const { option_id, value, customization } = selection;
            if (value) {
                if (customization.length > 0) customizable_options.push({ id: option_id, value_string: customization });
            } else {
                customizable_options.push({
                    id: option_id,
                    value_string: customization.map(c => c.id.toString()).join(',')
                });
            }
        });

        if (customizable_options.length > 0)
            return {
                customizable_options
            };
    }

    return {};
}
