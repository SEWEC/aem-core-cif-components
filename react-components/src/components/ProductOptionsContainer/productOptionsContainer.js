/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React, { useState, useEffect } from 'react';
import { useTranslation } from 'react-i18next';

import BundleProductOptions from './BundleProductOptions';
import CustomizableProductOptions from './CustomizableProductOptions';
import Button from '../Button';
import Price from '../Price';

import { useConfigContext } from '../../context/ConfigContext';
import * as dataLayerUtils from '../../utils/dataLayerUtils';
import * as optionsUtils from './utils';

const ProductOptionsContainer = () => {
    const [t] = useTranslation(['product', 'cart']);

    const [optionsState, setOptionsState] = useState({
        dataFetched: false,
        showCustomizeButton: false,
        productQuantity: 1,
        bundleState: null,
        customizableOptionsState: null
    });

    const { mountingPoints: { productOptionsContainer } } = useConfigContext();

    const sku = document.querySelector(productOptionsContainer)?.dataset?.sku
    const currencyCode = document.querySelector(productOptionsContainer)?.dataset?.currency;
    const rawProductData = document.querySelector(productOptionsContainer)?.dataset?.raw;
    const productId = document.querySelector('[data-cmp-is=product]')?.id;

    const productData = rawProductData ? JSON.parse(rawProductData)?.data?.products?.items[0] : null;

    useEffect(() => {
        if (productData) {
            let changes = { dataFetched: true };

            if (productData.__typename === "BundleProduct") {
                changes['bundleState'] = optionsUtils.createBundleState(productData.items);
                changes['showCustomizeButton'] = true;
            }

            if (productData?.options?.length > 0) {
                changes['customizableOptionsState'] = optionsUtils.createCustomizableOptionsState(productData.options);
            }

            setOptionsState({
                ...optionsState,
                ...changes
            })
        }
    }, []);

    const { dataFetched, showCustomizeButton, productQuantity, bundleState, customizableOptionsState } = optionsState;

    const getTotalPrice = () => {
        let price = 0;
        price += optionsUtils.getBundlePrice(bundleState);
        price += optionsUtils.getCustomizableOptionsPrice(customizableOptionsState);

        return <Price currencyCode={currencyCode} value={price} className="customPrice" />;
    }

    const changeQuantity = e => {
        const { target: { value } } = e;

        setOptionsState({
            ...optionsState,
            productQuantity: value
        });
    }

    const canAddToCart = () => {
        let result = optionsUtils.isOptionsStateValid(bundleState) && optionsUtils.isOptionsStateValid(customizableOptionsState);
        return result;
    }

    const addToCart = () => {
        const cartProduct = {
            sku,
            virtual: false,
            quantity: productQuantity,
            ...optionsUtils.getBundleCartOptions(bundleState),
            ...optionsUtils.getCustomizableCartOptions(customizableOptionsState)
        };

        const customEvent = new CustomEvent('aem.cif.add-to-cart', {
            detail: [cartProduct]
        });
        document.dispatchEvent(customEvent);
        // https://github.com/adobe/xdm/blob/master/docs/reference/datatypes/productlistitem.schema.md
        dataLayerUtils.pushEvent('cif:addToCart', {
            '@id': productId,
            'xdm:SKU': cartProduct.sku,
            'xdm:quantity': cartProduct.quantity,
            bundle: true
        });
    }

    if (!dataFetched) return <></>;

    if (showCustomizeButton) {
        return (
            <section className="productFullDetail__section productFullDetail__customizeBundle">
                <Button priority="high" onClick={() => setOptionsState({ ...optionsState, showCustomizeButton: false })}>
                    <span>{t('product:customize-bundle', 'Customize')}</span>
                </Button>
            </section>
        );
    }

    return (<>
        {bundleState !== null &&
            <BundleProductOptions
                key={"bundle-options"}
                currencyCode={currencyCode}
                data={bundleState}
                handleBundleChange={(newBundleState) => setOptionsState({ ...optionsState, bundleState: newBundleState })} />}
        {customizableOptionsState !== null &&
            <CustomizableProductOptions
                key={"customizable-options"}
                currencyCode={currencyCode}
                data={customizableOptionsState}
                handleOptionsChange={(newCustomizableOptionsState) => setOptionsState({ ...optionsState, customizableOptionsState: newCustomizableOptionsState })} />}
        <section className="productFullDetail__section productFullDetail__productOption">
            <h3>
                <span className="required">* {t('product:required-fields', 'Required fields')}</span>
                <span className="priceInfo">
                    {t('product:customization-price', 'Your customization')}: {getTotalPrice()}
                </span>
            </h3>
        </section>
        <section className="productFullDetail__quantity productFullDetail__section">
            <h2 className="productFullDetail__quantityTitle option__title">
                <span>{t('cart:quantity', 'Quantity')}</span>
            </h2>
            <div className="quantity__root">
                <span className="fieldIcons__root" style={{ '--iconsBefore': 0, '--iconsAfter': 1 }}>
                    <span className="fieldIcons__input">
                        <select
                            aria-label="product's quantity"
                            className="select__input field__input"
                            name="quantity"
                            value={productQuantity}
                            onChange={changeQuantity}>
                            <option value="1">1</option>
                            <option value="2">2</option>
                            <option value="3">3</option>
                            <option value="4">4</option>
                        </select>
                    </span>
                    <span className="fieldIcons__before"></span>
                    <span className="fieldIcons__after" >
                        <span className="icon__root" >
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                width="18"
                                height="18"
                                viewBox="0 0 24 24"
                                fill="none"
                                stroke="currentColor"
                                strokeWidth="2"
                                strokeLinecap="round"
                                strokeLinejoin="round" >
                                <polyline points="6 9 12 15 18 9"></polyline>
                            </svg>
                        </span>
                    </span>
                </span>
                <p className="message-root"></p>
            </div>
        </section>
        <section className="productFullDetail__cartActions productFullDetail__section" >
            <button
                className="button__root_highPriority button__root clickable__root button__filled"
                type="button"
                disabled={!canAddToCart()}
                onClick={addToCart} >
                <span className="button__content">
                    <span>{t('product:add-item', 'Add to Cart')}</span>
                </span>
            </button >
        </section >
    </>);
}

export default ProductOptionsContainer;
