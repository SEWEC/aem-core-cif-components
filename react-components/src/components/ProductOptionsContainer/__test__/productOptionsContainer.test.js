/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { render } from 'test-utils';
import { fireEvent, waitForDomChange } from '@testing-library/react';
import ProductOptionsContainer from '../productOptionsContainer';
import bundleData from './bundle-data.json';
import customizableOptions from './customizable-options.json';
import bundleCustomizableOptions from './bundle-with-customizable-options.json';

const config = {
    graphqlMethod: 'GET',
    storeView: 'default',
    graphqlEndpoint: 'endpoint',
    mountingPoints: {
        productOptionsContainer: '#product-options'
    }
};

describe('<ProductOptionsContainer>', () => {
    it('renders the component with no data', () => {
        const productOptionsContainer = document.createElement('div');
        productOptionsContainer.id = 'product-options';

        const { asFragment } = render(<ProductOptionsContainer />, {
            config: config,
            container: document.body.appendChild(productOptionsContainer)
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the component with bundle data', async () => {
        const productOptionsContainer = document.createElement('div');
        productOptionsContainer.id = 'product-options';
        productOptionsContainer.dataset.raw = JSON.stringify(bundleData);
        productOptionsContainer.dataset.sku = '24-WG080';
        productOptionsContainer.dataset.currency = 'USD';

        const { asFragment, container, getByRole } = render(<ProductOptionsContainer />, {
            config: config,
            container: document.body.appendChild(productOptionsContainer)
        });
        expect(asFragment()).toMatchSnapshot();

        const domChange = waitForDomChange({ container });
        fireEvent.click(getByRole('button', { name: 'Customize' }));

        await domChange;

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the component with customizable options', async () => {
        const productOptionsContainer = document.createElement('div');
        productOptionsContainer.id = 'product-options';
        productOptionsContainer.dataset.raw = JSON.stringify(customizableOptions);
        productOptionsContainer.dataset.sku = '24-UG06';
        productOptionsContainer.dataset.currency = 'USD';

        const { asFragment } = render(<ProductOptionsContainer />, {
            config: config,
            container: document.body.appendChild(productOptionsContainer)
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the component with customizable bundle data', async () => {
        const productOptionsContainer = document.createElement('div');
        productOptionsContainer.id = 'product-options';
        productOptionsContainer.dataset.raw = JSON.stringify(bundleCustomizableOptions);
        productOptionsContainer.dataset.sku = 'test-bundle';
        productOptionsContainer.dataset.currency = 'USD';

        const { asFragment, container, getByRole } = render(<ProductOptionsContainer />, {
            config: config,
            container: document.body.appendChild(productOptionsContainer)
        });
        expect(asFragment()).toMatchSnapshot();

        const domChange = waitForDomChange({ container });
        fireEvent.click(getByRole('button', { name: 'Customize' }));

        await domChange;

        expect(asFragment()).toMatchSnapshot();
    });
});
