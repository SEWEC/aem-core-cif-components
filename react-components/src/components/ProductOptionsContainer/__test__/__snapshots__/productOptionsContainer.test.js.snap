// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductOptionsContainer> renders the component with bundle data 1`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__customizeBundle"
  >
    <button
      class="root_highPriority"
      type="button"
    >
      <span
        class="content"
      >
        <span>
          Customize
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<ProductOptionsContainer> renders the component with bundle data 2`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Sprite Stasis Ball
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="product__options"
      >
        <label>
          <input
            checked=""
            name="1"
            type="radio"
            value="1"
          />
           Sprite Stasis Ball 55 cm +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__options"
      >
        <label>
          <input
            name="1"
            type="radio"
            value="2"
          />
           Sprite Stasis Ball 65 cm +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__options"
      >
        <label>
          <input
            name="1"
            type="radio"
            value="3"
          />
           Sprite Stasis Ball 75 cm +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <h2
        class="option__title productFullDetail__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Sprite Foam Yoga Brick
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="product__options"
      >
        <label>
          <input
            checked=""
            name="2"
            type="radio"
            value="4"
          />
           Sprite Foam Yoga Brick +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <h2
        class="option__title productFullDetail__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Sprite Yoga Strap
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="product__options"
      >
        <label>
          <input
            checked=""
            name="3"
            type="radio"
            value="5"
          />
           Sprite Yoga Strap 6 foot +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__options"
      >
        <label>
          <input
            name="3"
            type="radio"
            value="6"
          />
           Sprite Yoga Strap 8 foot +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__options"
      >
        <label>
          <input
            name="3"
            type="radio"
            value="7"
          />
           Sprite Yoga Strap 10 foot +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <h2
        class="option__title productFullDetail__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Sprite Foam Roller
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="product__options"
      >
        <label>
          <input
            checked=""
            name="4"
            type="radio"
            value="8"
          />
           Sprite Foam Roller +
          <b>
            <span
              class=""
            >
              $0.00
            </span>
          </b>
        </label>
      </div>
      <h2
        class="option__title productFullDetail__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3>
      <span
        class="required"
      >
        * Required fields
      </span>
      <span
        class="priceInfo"
      >
        Your customization: 
        <span
          class="customPrice"
        >
          $0.00
        </span>
      </span>
    </h3>
  </section>
  <section
    class="productFullDetail__quantity productFullDetail__section"
  >
    <h2
      class="productFullDetail__quantityTitle option__title"
    >
      <span>
        Quantity
      </span>
    </h2>
    <div
      class="quantity__root"
    >
      <span
        class="fieldIcons__root"
      >
        <span
          class="fieldIcons__input"
        >
          <select
            aria-label="product's quantity"
            class="select__input field__input"
            name="quantity"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
          </select>
        </span>
        <span
          class="fieldIcons__before"
        />
        <span
          class="fieldIcons__after"
        >
          <span
            class="icon__root"
          >
            <svg
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="message-root"
      />
    </div>
  </section>
  <section
    class="productFullDetail__cartActions productFullDetail__section"
  >
    <button
      class="button__root_highPriority button__root clickable__root button__filled"
      type="button"
    >
      <span
        class="button__content"
      >
        <span>
          Add to Cart
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<ProductOptionsContainer> renders the component with customizable bundle data 1`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__customizeBundle"
  >
    <button
      class="root_highPriority"
      type="button"
    >
      <span
        class="content"
      >
        <span>
          Customize
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<ProductOptionsContainer> renders the component with customizable bundle data 2`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Bag
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <span
          class="fieldIcons__root"
        >
          <span
            class="fieldIcons__input"
          >
            <select
              class="select__input field__input product__option"
            >
              <option
                value="9"
              >
                Joust Duffle Bag +$11.00
              </option>
              <option
                value="10"
              >
                Wayfarer Messenger Bag +$12.00
              </option>
            </select>
          </span>
          <span
            class="fieldIcons__before"
          />
          <span
            class="fieldIcons__after"
          >
            <span
              class="icon__root"
            >
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
      <h2
        class="product__quantityTitle"
      >
        <span>
          Quantity
        </span>
      </h2>
      <input
        class="option__quantity"
        min="1"
        type="number"
        value="1"
      />
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Backpack
      </span>
       
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <select
        class="product__option field__input"
        multiple=""
        name="6"
        size="5"
      >
        <option
          value="11"
        >
          Fusion Backpack +$0.00
        </option>
        <option
          value="12"
        >
          Driven Backpack +$0.00
        </option>
      </select>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Logo
      </span>
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="product__options"
      >
        <label>
          <input
            checked=""
            name="11"
            type="radio"
            value="15"
          />
           Yes +
          <b>
            <span
              class=""
            >
              $1.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__options"
      >
        <label>
          <input
            name="11"
            type="radio"
            value="16"
          />
           No +
          <b>
            <span
              class=""
            >
              $2.00
            </span>
          </b>
        </label>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Gift wrap
      </span>
    </h3>
    <div>
      <div
        class="product__option"
      >
        <label>
          <input
            type="checkbox"
            value="17"
          />
           White +
          <b>
            <span
              class=""
            >
              $1.00
            </span>
          </b>
        </label>
      </div>
      <div
        class="product__option"
      >
        <label>
          <input
            type="checkbox"
            value="18"
          />
           Black +
          <b>
            <span
              class=""
            >
              $2.00
            </span>
          </b>
        </label>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Custom message
      </span>
      <span>
         + 
        <span
          class=""
        >
          $3.00
        </span>
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <input
          class="field__input"
          value=""
        />
        <small>
          Maximum 200 characters
        </small>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Custom story
      </span>
      <span>
         + 
        <span
          class=""
        >
          $4.00
        </span>
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <textarea
          class="field__textarea"
          rows="5"
        />
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Dropdown option
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <span
          class="fieldIcons__root"
        >
          <span
            class="fieldIcons__input"
          >
            <select
              aria-label="Dropdown option"
              class="select__input field__input product__option"
              name="15"
            >
              <option
                value=""
              >
                None
              </option>
              <option
                value="19"
              >
                Op 1 +$1.00
              </option>
              <option
                value="20"
              >
                Op2 +$2.00
              </option>
            </select>
          </span>
          <span
            class="fieldIcons__before"
          />
          <span
            class="fieldIcons__after"
          >
            <span
              class="icon__root"
            >
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        MultiSelect
      </span>
    </h3>
    <div>
      <select
        class="product__option field__input"
        multiple=""
        name="16"
        size="5"
      >
        <option
          value="21"
        >
          Multi1 +$2.00
        </option>
        <option
          value="22"
        >
          Multi 2 +$3.00
        </option>
      </select>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3>
      <span
        class="required"
      >
        * Required fields
      </span>
      <span
        class="priceInfo"
      >
        Your customization: 
        <span
          class="customPrice"
        >
          $12.00
        </span>
      </span>
    </h3>
  </section>
  <section
    class="productFullDetail__quantity productFullDetail__section"
  >
    <h2
      class="productFullDetail__quantityTitle option__title"
    >
      <span>
        Quantity
      </span>
    </h2>
    <div
      class="quantity__root"
    >
      <span
        class="fieldIcons__root"
      >
        <span
          class="fieldIcons__input"
        >
          <select
            aria-label="product's quantity"
            class="select__input field__input"
            name="quantity"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
          </select>
        </span>
        <span
          class="fieldIcons__before"
        />
        <span
          class="fieldIcons__after"
        >
          <span
            class="icon__root"
          >
            <svg
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="message-root"
      />
    </div>
  </section>
  <section
    class="productFullDetail__cartActions productFullDetail__section"
  >
    <button
      class="button__root_highPriority button__root clickable__root button__filled"
      disabled=""
      type="button"
    >
      <span
        class="button__content"
      >
        <span>
          Add to Cart
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<ProductOptionsContainer> renders the component with customizable options 1`] = `
<DocumentFragment>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Drop Down Options
      </span>
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <span
          class="fieldIcons__root"
        >
          <span
            class="fieldIcons__input"
          >
            <select
              aria-label="Drop Down Options"
              class="select__input field__input product__option"
              name="1"
            >
              <option
                value="1"
              >
                value 1 +$20.00
              </option>
              <option
                value="2"
              >
                value 2 +$30.00
              </option>
            </select>
          </span>
          <span
            class="fieldIcons__before"
          />
          <span
            class="fieldIcons__after"
          >
            <span
              class="icon__root"
            >
              <svg
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Area Option
      </span>
      <span>
         + 
        <span
          class=""
        >
          $3.00
        </span>
      </span>
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <textarea
          class="field__textarea"
          rows="5"
        />
        <small>
          Maximum 200 characters
        </small>
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3
      class="option__title"
    >
      <span>
        Text Option
      </span>
      <span>
         + 
        <span
          class=""
        >
          $11.00
        </span>
      </span>
      <span
        class="required"
      >
         *
      </span>
    </h3>
    <div>
      <div
        class="productOptionSelect__root"
      >
        <input
          class="field__input"
          value=""
        />
      </div>
    </div>
  </section>
  <section
    class="productFullDetail__section productFullDetail__productOption"
  >
    <h3>
      <span
        class="required"
      >
        * Required fields
      </span>
      <span
        class="priceInfo"
      >
        Your customization: 
        <span
          class="customPrice"
        >
          $20.00
        </span>
      </span>
    </h3>
  </section>
  <section
    class="productFullDetail__quantity productFullDetail__section"
  >
    <h2
      class="productFullDetail__quantityTitle option__title"
    >
      <span>
        Quantity
      </span>
    </h2>
    <div
      class="quantity__root"
    >
      <span
        class="fieldIcons__root"
      >
        <span
          class="fieldIcons__input"
        >
          <select
            aria-label="product's quantity"
            class="select__input field__input"
            name="quantity"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
          </select>
        </span>
        <span
          class="fieldIcons__before"
        />
        <span
          class="fieldIcons__after"
        >
          <span
            class="icon__root"
          >
            <svg
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="message-root"
      />
    </div>
  </section>
  <section
    class="productFullDetail__cartActions productFullDetail__section"
  >
    <button
      class="button__root_highPriority button__root clickable__root button__filled"
      disabled=""
      type="button"
    >
      <span
        class="button__content"
      >
        <span>
          Add to Cart
        </span>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`<ProductOptionsContainer> renders the component with no data 1`] = `<DocumentFragment />`;
